### ISSUE TITLE
[configuration] Environment variable substitution syntax is hard to extend

### ISSUE THREAD

**Issue by mx-psi** [open-telemetry/opentelemetry-specification#3981]
The OTel SDK File configuration spec environment variable substitution is difficult to extend in a non-breaking way.

For example, adding the default syntax described on #3948 is a breaking change: currently the value must be left verbatim, while after this change it would be resolved to something. Since the spec is still experimental, breaking it is fine, but I think we should make the spec more resilient to this.

A way to approach this is to explicitly fail when a configuration file has  `${...}` expressions that have an invalid identifier/special characters. Turning a failure into a specific feature is not a breaking change and would allow us to extend things in the future if necessary.

If we fail on every invalid identifier, then we should have something like #3914, since there are valid cases for using `${...}` expressions in other contexts (e.g. on Prometheus relabel config using `${1}` is not unusual)

**Comment by jack-berg** [comment]
> currently the value must be left verbatim, while after this change it would be resolved to something

Can you elaborate on why this is breaking with an example of before and after?

**Comment by mx-psi** [comment]
Sure! So, my understanding of the spec is that taking this example and assuming `ENV` is unset:

```yaml
string_key: ${ENV:-default}
```

The resolved YAML before #3948 would be the same thing, since `S{ENV:-default}` does not match the regular expression defined on the spec:

```yaml
string_key: ${ENV:-default}  # type string
```

The resolved YAML after #3948 would be the `default` string:

```yaml
string_key: default  # type string
```

So the resolved configuration would be different on each of them. I would consider this a breaking change (one that people would rarely face but a breaking change nonetheless).

**Comment by TylerHelmuth** [comment]
+1 to failing when the configuration includes an invalid `${...}` expression. Repeating my thoughts from https://github.com/open-telemetry/opentelemetry-collector/pull/9854#discussion_r1554217814, I think failing fast is a clearer outcome for the user.

For users that mis-typed the env var name, keeping the string verbatim could mean an unintuitive error caused by the unwanted string. If we fail immediately we can log a clear error message that will be easier to troubleshoot.

If the user needs an illegal string like `${env:this-is-illegal}` they can use the escape to make it exact: `$${env:this-is-illegal}`.

To summarize: using the env var syntax is a request to parse an env var - if the provided env var cannot be parsed we should fail.

**Comment by jack-berg** [comment]
I think I misunderstood the goal. You're not saying to fail when an env var is referenced but unset. You're saying fail when an env var substitution expression doesn't quite match the regexp, and thus would be left as a string unless we fail.

**Comment by TylerHelmuth** [comment]
> You're saying fail when an env var substitution expression doesn't quite match the regexp, and thus would be left as a string unless we fail.

Ya this one

**Comment by jack-berg** [comment]
Yeah I agree that it makes sense to fail in this case.
